### ISSUE TITLE
Chinese Word Wrap is incorrect

### ISSUE THREAD

**Issue by kkbruce** [RickStrahl/MarkdownMonster#336]
If Enable Internal Preview Windows, Chinese Word Wrap is incorrect in Editor.

Try this (don't paste \`\`\` code, just pure Chinese words):

```
請看第 4 行，等號的右邊是一個 `Tuple` 物件。在等號的左邊，則使用了所謂的分解式宣告（deconstruction declaration）的語法。這行程式碼的作用是：把一個 `Tuple` 物件裡面的元素依序指派給等號左邊的括弧中宣告的變數。然後，我們就可以直接使用這些變數（倒數第 2 行）。
```

This is friend feedback https://www.youtube.com/watch?v=PuH9p7nTDfs&feature=youtu.be

**Comment by RickStrahl** [comment]
Hmmm... not really sure what I'm looking for here :-) But ultimately this ends up being an ACE Editor issue and it's probably better to submit that there for getting addressed. 

Can you try the following:

* Open the Preview (right click in Preview and Show In External Browser)
* What do you see in:
   * Chrome
   * Internet Explorer

It looks like the rendering in IE is wrapping funky but it does eventually wrap. But I suspect this has to do with the character spacing for since there are differing widths for the chinese and western characters.

IAC, I don't even know where to begin fixing this but it might be a good idea to post this on the ACE repository.

https://github.com/ajaxorg/ace/issues

With a repro case for Chrome (best) and/or IE (if it's different).

**Comment by kkbruce** [comment]
See [Word Wrap Problem](https://i.imgur.com/4HeHwA9.png)

I guess same #337 is IE Render problem. :-(

**Comment by huanlin** [comment]
hi @RickStrahl  and @kkbruce , 
I followed instructions on https://github.com/ajaxorg/ace and run an ACE editor locally, and I opened  http://localhost:8888/kitchen-sink.html with both IE 11 and Chrome, just to see if I can reproduce it.
And yes, I can reproduce it with IE 11, but Chrome is OK.
Specifically, I see the flickering (overlapping) issue while I'm typing Chinese in the ACE editor.

So yeah, it seems the issue happens only with IE + ACE + Chinese input method.
If ACE editor is not going to fix this issue, it seems the only workaround is to use another Web browser control from Markdown Monster. But that sounds a lot of work and I don't even sure whether it's a good idea or not.

By the way, thanks for the great work!

p.s.  I'm the guy who records and uploads the video: https://www.youtube.com/watch?v=PuH9p7nTDfs&feature=youtu.be

**Comment by RickStrahl** [comment]
The ACE Editor folks are pretty responsive if something doesn't work so there's a good chance it might get fixed. But - Chinese character set issues are often a difficult thing to debug, but if you can outline the issue to ACE at least we'll have an answer.

I've built a prototype browser addin that uses Chrome for Preview Rendering and was able to make that work. It was a bit of work but it's possible. The editor however has a lot more two way interaction so that's a much bigger job unfortunately. Additionally the Chrome plugin was considerably slower rendering than IE so running editor has been a tricky issue. I will pick this up in the future again when I have more time, but for now we're stuck with IE for rendering unfortunately.

**Comment by huanlin** [comment]
So I submit a ticket to ACE project: https://github.com/ajaxorg/ace/issues/3622
Hopefully there will be a fix in the near future.
Maybe it's time to close this issue now.
Thanks!

**Comment by RickStrahl** [comment]
Awesome - thank you. If you could update this issue that'd great. I'll leave this open a while longer.

**Comment by RickStrahl** [comment]
Ok so I just bungled an update on the Ace Forums because you're mixing issues in this issue with the IME overwrite issue. the link above links to the overwrite issue. This issue has not been posted to ACE I assume or is there another Issue you've filed for this one?

FWIW, I took a look to better understand what's going on with the wrapping issue. I also see this failing both in Markdown Monster (IE Hosted), standalone IE as well as Chrome. In all places I pretty much see the same behavior:

Here's inside of MM:

![image](https://user-images.githubusercontent.com/1374013/38215918-eb2dfb84-3664-11e8-80ad-e999ddbd905a.png)

For reference the preview shows what we probably should also be seeing in the editor in terms of wrapping.

Here's Chrome:

![image](https://user-images.githubusercontent.com/1374013/38216008-2abbf756-3665-11e8-9ce3-bb6dc9e2f1fd.png)

It seems that this is not isolated to IE...

We probably could use another issue in ACE for this one as well.

**Comment by RickStrahl** [comment]
Playing around with this a little more. If I take your text and inject some other Chinese characters into the document the wrapping seems to work (sorry for the random characters):

![image](https://user-images.githubusercontent.com/1374013/38245922-c00c0a6e-36db-11e8-8a18-bf6afca7fe6b.png)

Is there anything special about the string you have there? Why the difference?

**Comment by huanlin** [comment]
Oops, the ticket I submitted to ACE is only about overwriting issue. I didn't submit a ticket about the wrapping issue.

As to wrapping,  I'm not sure if there are anything special in the string. I thought it might be similar to the issue I've seen from another library: [asciidoctor-pdf](https://github.com/asciidoctor/asciidoctor-pdf). 

After I found some Chinese specific issues in asciidocor-pdf, I gave it up. But I had tried to see what's going with the wrapping issue, and I found a patch from a Chinese guy. Here is the patch written in Ruby: 
https://github.com/chloerei/asciidoctor-pdf-cjk/blob/master/lib/asciidoctor/pdf/cjk/converter.rb
and here:
https://github.com/chloerei/asciidoctor-pdf-cjk/blob/master/lib/asciidoctor/pdf/cjk.rb

I don't speak Ruby, so honestly I am not sure if it is relevant. My investigation stopped there.
